### PR TITLE
Reading data

### PR DESCRIPTION
**Pull request by NathanKelly-ONS**
Merge request template: please remove the appropriate parts of this template. 

Pre-merge request checklist (to be completed by the one making the request):
 - [x] I have performed a full review of this code myself.
    - For Python code in PySpark specific sections, all code should have been run in Jupyter notebooks. 
    - For code in sections of the book containing both Python and R code, the page of the book should be constructed as described in the [contributing guide](https://github.com/best-practice-and-impact/ons-spark/blob/main/CONTRIBUTING.md) and converted to a markdown file. 
 - [x] I have formatted the outputs of code blocks correctly (to match other outputs in the book and in line with the style guide [coming soon])
 - [x] I have built the book as outlined in the [contributing guide](https://github.com/best-practice-and-impact/ons-spark/blob/main/CONTRIBUTING.md) and confirmed that any additional/modified content is displaying as expected.

Details of this request (such as):
Creating a new chapter on reading and writing data in Spark.

Things to note about this request (such as):
Wasn't able to get the SparklyR Avro to work, so will create an issue for that.

Requirements for review (such as):
Review content, build the book and ensure it's displaying correctly.

### CODE REVIEW

**Comment by NathanKelly-ONS** [comment]
@robertswh having trouble getting the notebook converter to work so have just created this as a draft PR for now (since I haven't been able to convert the notebooks/build the book) but would be good to get your feedback on the content and I'll add in the markdown and check the book once I can get the converter to work.

File: ons-spark/pyspark-intro/reading-data-pyspark.ipynb [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+    "\n",
+    "animal_rescue = spark.read.csv(path = config[\"rescue_path_csv\"], header = True, schema = custom_schema)\n",
+    "animal_rescue.printSchema()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can see that DateTimeOfCall has now been read in correctly as a timestamp type. Also, providing a schema will improve the efficiency of the operation. This is because, in order to infer a schema, Spark needs to scan the dataset. A column could contain, for example, an integer for the first 1000 rows and a string for the 1001th row - in which case it would be inferred as a string, but this wouldn't be obvious from the first 1000 rows. Needing to sample the data in each column can be quite memory intensive. Providing a schema means that Spark no longer has to sample the data before reading it in. "
```
Comment (robertswh) [inline ons-spark/pyspark-intro/reading-data-pyspark.ipynb:148]: Here it is! I hadn't thought about the point that scanning is memory intensive, I suppose that is true. What I remember from my training is that it takes time and will be repeated with repeated actions, but fair point about resource. It's just unnecessary resource being used isn't it.

File: ons-spark/pyspark-intro/reading-data-pyspark.ipynb [hunk 2]
```diff
@@ -0,0 +1,358 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Reading Data in PySpark\n",
+    "This chapter will go into more detail about the various methods available to read in data in PySpark. The previous chapter, [Introduction to PySpark](../pyspark-intro/pyspark-intro), briefly covered [`spark.read.csv()`](https://spark.apache.org/docs/latest/sql-data-sources-csv.html) and [`spark.read.parquet()`](https://spark.apache.org/docs/latest/sql-data-sources-parquet.html) to read data in as a CSV and parquet file respectively.\n",
```
Comment (robertswh) [inline ons-spark/pyspark-intro/reading-data-pyspark.ipynb:8]: I think the chapter is less about various methods to read in data, and more about how Spark interacts with the different file formats. Does that make sense?

**Review by robertswh — COMMENTED** [review]
I think this is great. Few minor points but you're definitely on the right track I think. We might want to make changes once we understand how we can store data on CDP and GCP, but we have to start somewhere.

File: ons-spark/pyspark-intro/reading-data-pyspark.ipynb [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "It's important to note the two arguments we have provided to the [`spark.read.csv()`](https://spark.apache.org/docs/latest/sql-data-sources-csv.html) function, `header` and `inferSchema`.\n",
+    "\n",
+    "By setting `header` to True, we're saying that we want the top row to be used as the column names. If we did not set this argument to True, then the top rows will be treated as the first row of data, and columns will be given a default name of \"_c1\", \"_c2\", \"_c3\" and so on.\n",
+    "\n",
+    "`inferSchema` is very important - a disadvantage of using a CSV file is that they are not associated with a schema in the same way parquet files are. By setting `inferSchema` to True, we're allowing the PySpark API to attempt to work out the schemas based on the contents of each column. If this were set to False, then each column would be set to a string datatype by default.\n",
```
[the remark this answers is not in the document]
Comment (NathanKelly-ONS) [reply ons-spark/pyspark-intro/reading-data-pyspark.ipynb:57]: @robertswh are you happy that this has been spoken about in your next comment below this one?

File: ons-spark/pyspark-intro/reading-data-pyspark.ipynb (hunk 1 above)
Comment (NathanKelly-ONS) [reply ons-spark/pyspark-intro/reading-data-pyspark.ipynb:148]: Added in a section about it taking longer and being repeated every action.

File: ons-spark/pyspark-intro/reading-data-pyspark.ipynb [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+   "source": [
+    "In many ways, Hive tables have a lot of the same benefits that a parquet file has, such as the storing of schemas and supporting predicate pushdowns. In fact, a Hive table may consist of parquet files - a parquet file is the default underlying file structure PySpark uses when saving out a Hive table. A Hive table can consist of any of the underlying data formats we've discussed in this chapter.\n",
+    "\n",
+    "The key benefits that a Hive table offers over other file formats are:\n",
+    "\n",
+    "* Metadata - Hive tables contain more detailed metadata than parquet files, though both store information on schemas.\n",
+    "* Mutability - Hive tables are mutable, this means you can modify them in place using the [ALTER TABLE](https://spark.apache.org/docs/latest/sql-ref-syntax-ddl-alter-table.html#:~:text=SET%20TABLE%20PROPERTIES,to%20drop%20the%20table%20property.) SQL statement. This can be used to add, modify or delete columns in a table. In contrast, parquet files are immutable - to make a change to a parquet file, you have to read it in first, make the changes and then save it by either overwriting the original file or by creating a new file."
```
[the remark this answers is not in the document]
Comment (NathanKelly-ONS) [reply ons-spark/pyspark-intro/reading-data-pyspark.ipynb:287]: I will remove the section about in-place then. Can you clarify what you mean by "you can't modify by row, you can by partition but not by row" - as in, you can only modify Hive tables by partition but not by row? What happens under the hood if you use SQL to insert a row into the HIVE table then?

File: ons-spark/raw-notebooks/reading-and-writing-data/reading-and-writing-data-spark.ipynb [hunk 5]
```diff
@@ -0,0 +1,609 @@
...
+   "metadata": {},
+   "source": [
+    "```r\n",
+    "library(sparklyr)\n",
+    "library(dplyr)\n",
+    "\n",
+    "sc <- sparklyr::spark_connect(\n",
+    "  master = \"local[2]\",\n",
+    "  app_name = \"ons-spark\",\n",
+    "  config = list(default = list(spark.submit.deployMode = \"cluster\")),\n",
```
Comment (ChrisSoderberg-ONS) [inline ons-spark/raw-notebooks/reading-and-writing-data/reading-and-writing-data-spark.ipynb:43]: Can't remember why I set the config parameter on spark_connect like this but probably better to change it to config = sparklyr::spark_config() to be consistent with the rest of the book

**Review by ChrisSoderberg-ONS — COMMENTED** [review]
I would just like to change the config parameter in the spark_connect code example but other than that everything looks fine to me

File: ons-spark/raw-notebooks/reading-and-writing-data/reading-and-writing-data-spark.ipynb (hunk 5 above)
Comment (NathanKelly-ONS) [reply ons-spark/raw-notebooks/reading-and-writing-data/reading-and-writing-data-spark.ipynb]: Good catch, I've amended that and checked and the book is displaying correctly with those changes.

**Review by ChrisSoderberg-ONS — APPROVED** [review]

**Pull request merged**